plugins/auth-backend/src/providers/bitbucketServer/provider.ts

```diff
@@ -0,0 +1,304 @@
+    this.signInResolver = options.signInResolver;
+    this.authHandler = options.authHandler;
+    this.resolverContext = options.resolverContext;
+    this.strategy = new OAuth2Strategy(
+      {
+        authorizationURL: options.authorizationUrl,
+        tokenURL: options.tokenUrl,
+        clientID: options.clientId,
+        clientSecret: options.clientSecret,
+        callbackURL: options.callbackUrl,
+      },
+      (
+        accessToken: string,
+        refreshToken: string,
+        params: any,
+        fullProfile: PassportProfile,
+        done: VerifyCallback,
+      ) => {
+        done(undefined, { fullProfile, params, accessToken }, { refreshToken });
+      },
+    );
+    this.host = options.host;
+  }
+
+  async start(req: OAuthStartRequest): Promise<OAuthStartResponse> {
+    return await executeRedirectStrategy(req, this.strategy, {
+      accessType: 'offline',
+      prompt: 'consent',
+      scope: req.scope,
+      state: encodeState(req.state),
+    });
+  }
+
+  async handler(
+    req: express.Request,
+  ): Promise<{ response: OAuthResponse; refreshToken?: string }> {
+    const { result, privateInfo } = await executeFrameHandlerStrategy<
+      BitbucketServerOAuthResult,
+      PrivateInfo
+    >(req, this.strategy);
+
+    return {
+      response: await this.handleResult(result),
+      refreshToken: privateInfo.refreshToken,
+    };
+  }
+
+  async refresh(
+    req: OAuthRefreshRequest,
+  ): Promise<{ response: OAuthResponse; refreshToken?: string }> {
+    const { accessToken, refreshToken, params } =
+      await executeRefreshTokenStrategy(
+        this.strategy,
+        req.refreshToken,
+        req.scope,
+      );
+    const fullProfile = await executeFetchUserProfileStrategy(
+      this.strategy,
+      accessToken,
+    );
+    return {
+      response: await this.handleResult({
+        fullProfile,
+        params,
+        accessToken,
+      }),
+      refreshToken,
+    };
+  }
+
+  private async handleResult(
+    result: BitbucketServerOAuthResult,
+  ): Promise<OAuthResponse> {
+    // The OAuth2 strategy does not return a user profile -> let's fetch it before calling the auth handler
+    result.fullProfile = await this.fetchProfile(result);
+    const { profile } = await this.authHandler(result, this.resolverContext);
+
+    let backstageIdentity = undefined;
+    if (this.signInResolver) {
+      backstageIdentity = await this.signInResolver(
+        { result, profile },
+        this.resolverContext,
+      );
+    }
+
+    return {
+      providerInfo: {
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
+      },
+      profile,
+      backstageIdentity,
+    };
+  }
+
+  private async fetchProfile(
+    result: BitbucketServerOAuthResult,
+  ): Promise<PassportProfile> {
+    // Get current user name
+    let whoAmIResponse;
+    try {
+      whoAmIResponse = await fetch(
```

Comment:
You'll want to `import fetch from 'node-fetch'`, it's not available everywhere yet

Reply:
Fixed - thanks for your help! :) 
As a person who is doing mostly devops and backend stuff, those things are quite hard for me to get started with